src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -426,45 +426,53 @@ public static string SchemeToString(HttpScheme scheme)
             }
         }
 
-        public static bool IsValidHostHeader(string hostText)
+        public static void ValidateHostHeader(string hostText)
         {
-            // The spec allows empty values
-            if (string.IsNullOrEmpty(hostText))
+            // This is a string.IsNullOrEmpty test, but arranged to elmininate the
+            // bounds check from accessing the firstChar of the string
+            if (hostText is null || 0u >= (uint)hostText.Length)
             {
-                return true;
+                // The spec allows empty values
+                return;
             }
 
-            if (hostText[0] == '[')
+            var firstChar = hostText[0];
+            if (firstChar == '[')
             {
-                return IsValidIPv6Host(hostText);
+                // Tail call
+                ValidateIPv6Host(hostText);
             }
-
-            if (hostText[0] == ':')
+            else
             {
-                // Only a port
-                return false;
-            }
+                if (firstChar == ':')
+                {
+                    // Only a port
+                    BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
+                }
 
-            var i = 0;
-            for (; i < hostText.Length; i++)
-            {
-                if (!IsValidHostChar(hostText[i]))
+                // Enregister array
+                var hostCharValidity = HostCharValidity;
+                var i = 0;
+                for (; i < hostText.Length; i++)
                 {
-                    break;
+                    var ch = (int)hostText[i];
+                    // Bounds check and elimiate second bounds check
+                    if ((uint)ch >= (uint)hostCharValidity.Length || !hostCharValidity[ch])
+                    {
+                        break;
+                    }
                 }
-            }
-            return IsValidHostPort(hostText, i);
-        }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static bool IsValidHostChar(char ch)
-        {
-            return ch < HostCharValidity.Length && HostCharValidity[ch];
+                if (i < hostText.Length)
+                {
+                    // Tail call
+                    ValidateHostPort(hostText, i); 
+                }
+            }
         }
 
         // The lead '[' was already checked
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Are you removing this attribute from all the methods you expect to be TCO'd?

Reply:
Ish, making sure they are tail called (easy way, by not returning anything, no funky params and last action in function)

But mostly its not inlining where there is no overall fast-path. So origin form is fully inlined, as a the char checks; but what's the right choice for the others? Is ipv6 via raw ip address a common scenerio? Is ports a common scenerio? (might be due to https on 443; but isn't for http on 80)